aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/SynchronousNonBlockingStepTest.java

```diff
@@ -0,0 +1,245 @@
+        SynchronousNonBlockingStep.waitForStart("wait", b);
+
+        // At this point the execution is paused inside the synchronous non-blocking step
+        // Check for FlowNode created
+        FlowGraphWalker walker = new FlowGraphWalker(b.getExecution());
+        boolean found = false;
+        for (FlowNode n = walker.next(); n != null; n = walker.next()) {
+            if (n instanceof StepNode) {
+                found = true;
+                break;
+            }
+        }
+
+        System.out.println("Checking flow node added...");
+        assertTrue("FlowNode has to be added just when the step starts running", found);
+
+        // Check for message the test message sent to context listener
+        System.out.println("Checking build log message present...");
+        j.waitForMessage("Test Sync Message", b);
+        // The last step did not run yet
+        j.assertLogNotContains("Second message", b);
+
+        // Let syncnonblocking to continue
+        SynchronousNonBlockingStep.notify("wait");
+
+        System.out.println("Waiting until syncnonblocking (and the full flow) finishes");
+        j.waitForCompletion(b);
+        System.out.println("Build finished. Continue.");
+        // Check for the last message
+        j.assertLogContains("Second message", b);
+        j.assertBuildStatusSuccess(b);
+    }
+
+    @Test
+    public void interruptedTest() throws Exception {
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("node {\n" +
+            "echo 'First message'\n" +
+            "syncnonblocking 'waitinterrupt'\n" +
+            "echo 'Second message'\n" +
+        "}"));
+        WorkflowRun b = p.scheduleBuild2(0).getStartCondition().get();
+
+        // Wait for syncnonblocking to be started
+        System.out.println("Waiting to syncnonblocking to start...");
+        // waitinterrupt is a special key that makes syncnonblocking step to wait inside a while  loop checking
+        // for Thread.interrupted()
+        SynchronousNonBlockingStep.waitForStart("waitinterrupt", b);
+
+        // At this point syncnonblocking is looping and checking for interruption
+
+        CpsFlowExecution e = (CpsFlowExecution) b.getExecutionPromise().get();
+        // Let's force a call to stop. This will try to send an interruption to the run Thread
+        e.interrupt(Result.ABORTED);
+        System.out.println("Looking for interruption received log message");
+        j.waitForMessage("Run thread interrupted", b);
+        j.waitForCompletion(b);
+    }
+
+    @Test
+    public void parallelTest() throws Exception {
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("node {\n" +
+            "echo 'First message'\n" +
+            "parallel( a: { syncnonblocking 'wait0'; echo 'a branch'; }, b: { echo 'b branch' } )\n" +
```

Comment:
:ant: this does not prove anything, because you are not preventing the second `echo` from running before `syncnonblocking`. You would need to use `semaphore` in the second branch so that you could wait for the `syncnonblocking` step to start but sleep, _then_ run the `echo 'b branch'` and demonstrate that the message was sent to the build log, then resume.